### ISSUE TITLE
Use of g_autofree and g_autoptr

### ISSUE THREAD

**Issue by whot** [linuxwacom/libwacom#142]
glib has two useful macros `g_autofree` and `g_autoptr` that we could use in our code. They're wrappers around `__attribute__((__cleanup__))` and make things a bit more straightforward:

```
int foo(void) {
   g_autofree char *name = NULL;
   name = strdup("bar");
   return 1;
}
```
The `g_autofree` macro tells the compiler to add the required `g_free(name)` before the return statement. The `g_autoptr` macro is effectively the same but for `GObject`.

Using those would make the code slightly easier in some places, but comes with restrictions:
- they're only supported on gcc, clang and (I think) icc but not MSVC. Not sure we care about that bit anyway.
- our own structs aren't GObjects so we cannot use them, we'd have to define our own cleaup functions. Which is trivial though.

Any comments?

**Comment by jigpu** [comment]
That's a cute trick, but its going to take a while for my inner proofreader to get used to it. Limiting to GCC and Clang should be fine; one of those should be the default compiler on most any Linux or BSD system out there (but maybe(?) not Solaris, if we care :man_shrugging: ).

It looks like the automatic cleanup has [its own footguns](https://blog.fishsoup.net/2015/11/05/attributecleanup-mixed-declarations-and-code-and-goto/) to be avoided, but only their `count1()` example is both buggy and idiomatic. Probably nothing to worry about.

**Comment by whot** [comment]
Simplest rule is "always initialize to NULL" and you're pretty much good to go. Alas, and this is the second time this bites me (we have problems like this in libratbag as well): clang analyzer has random false positives when you start using ``__attribute__((__cleanup__)``. And it does not have a way to shut up specific warnings, so you're left with them until you can fix them.

Here's my branch with the conversions:
https://github.com/whot/libwacom/tree/wip/autoptr

One of the false positives is `bus_str` in `get_device_info()` in `libwacom/libwacom.c`. It took me a lot of trial and error and random guesses to figure out how to make that one disappear. 

The other two were even worse: one was changing from `realloc` to `reallocarray`, the second one from `strdup` to `g_strdup`. Functionally equivalent but one caused false positives, the other one doesn't.

So yeah, the issue with `g_autoptr` and related is that we'll spend more time trying to figure out how to work around clang-analyzer's false positives than we'll save using it in the first place. Right now we're at less than 30 min to switch everything, 3 hours trying things to make c-a shut up.

**Comment by jigpu** [comment]
Yikes. I'd probably wait for clang's implementation to more fully bake in that case.

**Comment by whot** [comment]
Yeah, I agree. This was more than frustrating and the little benefits we get from this isn't worth the hours of debugging.

**Comment by whot** [comment]
Side note 5 years later - this is still an issue and we're still getting the same false positives with `clang version 18.1.6 (Fedora 18.1.6-3.fc40)`

**Comment by whot** [comment]
If we shift from `scan-build` to use `clang-tidy` (available as meson target if `.clang-tidy` is present) then we can disable these warnings with:

```
$ cat .clang-tidy
Checks: '-clang-analyzer-unix.Malloc'
```
Frustratingly, this also disables [double free and use-after-free](https://clang.llvm.org/docs/analyzer/checkers.html#unix-malloc) but with consistent use of the attribute these shouldn't be an issue.

Of course it also shows up other warnings we'll have to fix first.
